service/raft/raft_group0.cc

```diff
@@ -587,6 +739,503 @@ persistent_discovery::persistent_discovery(raft::server_address self, const peer
+    } else {
+        rslog.info(
+            "upgrade_to_group0(): we're already a member of group 0."
+            " Apparently we're restarting after a previous upgrade attempt failed.");
+    }
+
+    // After we joined, we shouldn't be removed from group 0 until the end of the procedure.
+    // The implementation of `leave_group0` waits until upgrade finishes before leaving the group.
+    // There is no guarantee that `remove_from_group0` from another node (that has
+    // finished upgrading) won't remove us after we enter `synchronize` but before we leave it;
+    // but then we're not needed for anything anymore and we can be shutdown,
+    // and we won't do anything harmful to other nodes while in `synchronize`, worst case being
+    // that we get stuck.
+
+    auto& group0_server = _raft_gr.group0();
+
+    rslog.info("upgrade_to_group0(): waiting until every peer has joined Raft group 0...");
+    co_await wait_until_every_peer_joined_group0(sys_ks, group0_server, _abort_source);
+    rslog.info("upgrade_to_group0(): every peer is a member of Raft group 0.");
+
+    if (start_state == group0_upgrade_state::use_old_procedures) {
+        // We perform a schema synchronization step before entering `synchronize` upgrade state.
+        //
+        // This step is not necessary for correctness: we will make sure schema is synchronized
+        // after every node enters `synchronize`, where schema changes are disabled.
+        //
+        // However, it's good for reducing the chance that we get stuck later. If we manage to ensure that schema
+        // is synchronized now, there's a high chance that after entering `synchronize` state we won't have
+        // to do any additional schema pulls (only verify quickly that the schema is still in sync).
+        rslog.info("upgrade_to_group0(): waiting for schema to synchronize across all nodes in group 0...");
+        auto can_finish_early = [] { return make_ready_future<bool>(false); };
+        co_await synchronize_schema(_qp.db().real_database(), _ms, group0_server, _mm, can_finish_early, _abort_source);
+
+        // Before entering `synchronize`, perform a round-trip of `get_group0_upgrade_state` RPC calls
+        // to everyone as a dry run, just to check that nodes respond to this RPC.
+        // Obviously we may lose connectivity immediately after this function finishes,
+        // causing later steps to fail, but if network/RPC module is already broken, better to detect
+        // it now than after entering `synchronize` state. And if this steps succeeds, then there's
+        // a very high chance that the following steps succeed as well (we would need to be very unlucky otherwise).
+        rslog.info("upgrade_to_group0(): perfoming a dry run of remote `get_group0_upgrade_state` calls...");
+        co_await check_remote_group0_upgrade_state_dry_run(group0_server, _ms, _abort_source);
```

Comment:
I am not sure why this is needed. We need to handle the failure anyway.

Reply:
It's not needed for correctness, but I included this step to improve liveness in case we're not able to synchronize schema.

Suppose we're not able to synchronize schema (due to network connectivity error or node down or whatever) and we don't have this step. Then:
- we will enter `synchronize` state
- we will then block on trying to synchronize it
- schema changes are unavailable due to `synchronize` state

On the other hand with this step:
- we will block on trying to synchronize schema
- we don't enter `synchronize` step
- upgrade is stuck, but schema changes are still available

So generally it's harder to lose schema change availability if we have this step